### PR TITLE
chore: Upgrade libp2p and bitswap

### PR DESCRIPTION
**Pull request by hanabi1224**
**Summary of changes**
Changes introduced in this pull request:
- [x] Upgrade `libp2p` to `0.49.0`
- [x] Upgrade `libp2p-bitswap` to `0.24`
- [x] Fix and verify new `bitswap` logic
- [x] Install `protoc` in CI env and `Dockerfile` which is required by [`libp2p 0.47+`](https://github.com/libp2p/rust-libp2p/releases/tag/v0.47.0)  It reduces build time by not building `protoc` from source)
- [x] Remove `tiny-cid`
- [x] Fix build on MacOS
- [x] Improve prometheus metrics by splitting inbound/outbound request/response
- [x] Verify metrics at http://localhost:6116/metrics
- [x] ~~Manual test on calibnet~~ `calibnet sync check` CI job passes
- [x] Manual test on mainnet

**Reference issue to close (if applicable)**

Closes https://github.com/ChainSafe/forest/issues/1924 https://github.com/ChainSafe/forest/issues/1920

**Other information and links**

### CODE REVIEW

File: node/forest_libp2p/src/service.rs
Comment (creativcoder) [inline node/forest_libp2p/src/service.rs:520]: nit: can remove this

File: node/forest_libp2p/src/service.rs
Comment (creativcoder) [inline node/forest_libp2p/src/service.rs:635]: Would be good to log this too for debugging purposes in case this events gets hit

File: node/forest_libp2p/src/service.rs
Comment (hanabi1224) [reply node/forest_libp2p/src/service.rs]: fixed

File: node/forest_libp2p/src/service.rs
Comment (hanabi1224) [reply node/forest_libp2p/src/service.rs:651]: Not sure if it will flood the log, original logic does not log here

File: node/db/src/rocks.rs
Comment (lemmih) [inline node/db/src/rocks.rs:157]: Mark code blocks to let the spellchecker know that these aren't just weirdly spelled words.

```suggestion
    /// `fvm_ipld_encoding::DAG_CBOR(0x71)` is covered by [`libipld::DefaultParams`]
```

File: .config/forest.dic
Comment (lemmih) [inline .config/forest.dic:24]: This is code and not English words. They belong in code blocks rather than the dictionary.

**Review by lemmih — CHANGES_REQUESTED** [review]

File: node/db/src/rocks.rs
Comment (hanabi1224) [reply node/db/src/rocks.rs]: fixed

File: .config/forest.dic
Comment (hanabi1224) [reply .config/forest.dic]: fixed

File: node/forest_libp2p/src/service.rs
Comment (hanabi1224) [inline node/forest_libp2p/src/service.rs:554]: This should be the only thing that I cannot find a way to convert, to my understanding this is only for instrumentation. Any suggestions? @lemmih

File: Cargo.lock
Comment (hanabi1224) [inline Cargo.lock:8495]: tiny-cid is removed

File: node/forest_libp2p/src/service.rs
Comment (hanabi1224) [reply node/forest_libp2p/src/service.rs]: nvm this is fixed

**Comment by lemmih** [comment]
Running a manual test on a droplet. Will update when it's done.

File: Cargo.lock
Comment (tyshko5) [reply Cargo.lock:8592]: Thanks!

**Comment by lemmih** [comment]
This branch work for connecting to calibnet but I'm having difficulties with mainnet. Seeing a lot of this error:
>  message hasn't set `send_dont_have`: skipping

**Comment by lemmih** [comment]
Hmm, the DO droplet is too slow to match up to the most recent tipset. But it hasn't failed so far. Just a lots of notices that seem to be harmless.

**Comment by hanabi1224** [comment]
> Can we do something about the error messages and warnings? 

Sure, let me confirm the error message does not mean `libp2p-bitswap` not working and I will set its log level to `fatal`

**Comment by lemmih** [comment]
I'm completely synced to mainnet and everything appears to be in order! Nice!

**Review by lemmih — CHANGES_REQUESTED** [review]
Works great. But let's use `protobuf-compiler` from Ubuntu and let's update the build instructions for ubuntu and fedora.

File: Cargo.toml
Comment (lemmih) [inline Cargo.toml:114]: Should we push your branch to the ChainSafe bitswap repository?

File: README.md
Comment (lemmih) [inline README.md:64]: You can add `protobuf-compiler` to the Ubuntu and Fedora steps. Not sure what to do on Archlinux.

File: scripts/install_protoc_linux.sh
Comment (lemmih) [inline scripts/install_protoc_linux.sh:7]: This script can be deleted, right?

File: Cargo.toml
Comment (hanabi1224) [reply Cargo.toml:114]: That will be great, I don't think I have write access, would u push it and let me know?

File: scripts/install_protoc_linux.sh
Comment (hanabi1224) [reply scripts/install_protoc_linux.sh]: Fixed

File: README.md
Comment (hanabi1224) [reply README.md:64]: Fixed

File: Cargo.toml
Comment (lemmih) [reply Cargo.toml:114]: I can merge a PR to that repository. Let's do it after this PR has been merged.

**Review by lemmih — APPROVED** [review]

File: Cargo.toml
Comment (hanabi1224) [reply Cargo.toml:114]: Sounds good

File: node/db/src/rocks.rs
Comment (LesnyRumcajs) [inline node/db/src/rocks.rs:180]: Can we extract this method? Seems it will be exactly the same for all structs implementing `Store`.

File: node/forest_libp2p/src/service.rs
Comment (LesnyRumcajs) [inline node/forest_libp2p/src/service.rs:404]: It would be nice to refactor such oversized methods so there's a single `match` calling methods. 300 LoC for a method isn't ideal.

**Review by LesnyRumcajs — APPROVED** [review]
In general, looks quite solid!

File: node/db/src/rocks.rs
Comment (hanabi1224) [reply node/db/src/rocks.rs:181]: Good suggestion, fixed

**Review by elmattic — APPROVED** [review]

**Pull request merged**
